src/combinatorics/young_diagram.lean

```diff
@@ -0,0 +1,127 @@
+standard order on products. -/
+@[ext] structure young_diagram :=
+(cells : finset (ℕ × ℕ))
+(is_lower_set : is_lower_set (cells : set (ℕ × ℕ)))
+
+namespace young_diagram
+
+instance : set_like young_diagram (ℕ × ℕ) :=
+{ coe            := coe young_diagram.cells,
+  coe_injective' := λ μ ν h, by { rwa [young_diagram.ext_iff, ← finset.coe_inj] } }
+
+@[simp] lemma mem_cells {μ : young_diagram} (c : ℕ × ℕ) :
+  c ∈ μ.cells ↔ c ∈ μ := iff.rfl
+
+/-- In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+    means (i1, j1) is weakly up-and-left of (i2, j2). -/
+lemma up_left_mem (μ : young_diagram) {i1 i2 j1 j2 : ℕ}
+  (hi : i1 ≤ i2) (hj : j1 ≤ j2) (hcell : (i2, j2) ∈ μ) : (i1, j1) ∈ μ :=
+μ.is_lower_set (prod.mk_le_mk.mpr ⟨hi, hj⟩) hcell
+
+section distrib_lattice
+
+@[simp] lemma cells_subset_iff {μ ν : young_diagram} : μ.cells ⊆ ν.cells ↔ μ ≤ ν := iff.rfl
+@[simp] lemma cells_ssubset_iff {μ ν : young_diagram} : μ.cells ⊂ ν.cells ↔ μ < ν := iff.rfl
+
+instance : has_sup young_diagram :=
+{ sup := λ μ ν, { cells        := μ.cells ∪ ν.cells,
+                  is_lower_set := by { rw finset.coe_union,
+                                       exact μ.is_lower_set.union ν.is_lower_set } } }
+
+@[simp] lemma cells_sup (μ ν : young_diagram) : (μ ⊔ ν).cells = μ.cells ∪ ν.cells := rfl
+
+@[simp, norm_cast] lemma coe_sup (μ ν : young_diagram) : ↑(μ ⊔ ν) = (μ ∪ ν : set (ℕ × ℕ)) :=
+finset.coe_union _ _
+
+@[simp] lemma mem_sup {μ ν : young_diagram} {x : ℕ × ℕ} : x ∈ (μ ⊔ ν) ↔ x ∈ μ ∨ x ∈ ν :=
+finset.mem_union
+
+instance : has_inf young_diagram :=
+{ inf := λ μ ν, { cells        := μ.cells ∩ ν.cells,
+                  is_lower_set := by { rw finset.coe_inter,
+                                       exact μ.is_lower_set.inter ν.is_lower_set } } }
+
+@[simp] lemma cells_inf (μ ν : young_diagram) : (μ ⊓ ν).cells = μ.cells ∩ ν.cells := rfl
+
+@[simp, norm_cast] lemma coe_inf (μ ν : young_diagram) : ↑(μ ⊓ ν) = (μ ∩ ν : set (ℕ × ℕ)) :=
+finset.coe_inter _ _
+
+@[simp] lemma mem_inf {μ ν : young_diagram} {x : ℕ × ℕ} : x ∈ (μ ⊓ ν) ↔ x ∈ μ ∧ x ∈ ν :=
+finset.mem_inter
+
+/-- The empty Young diagram is (⊥ : young_diagram). -/
+instance : order_bot young_diagram :=
+{ bot := { cells := ∅, is_lower_set := λ _ _ _, false.elim }, bot_le := λ _ _, false.elim }
+
+@[simp] lemma cells_bot : (⊥ : young_diagram).cells = ∅ := rfl
+
+@[simp, norm_cast] lemma coe_bot : ↑(⊥ : young_diagram) = (∅ : set (ℕ × ℕ)) := rfl
+
+@[simp] lemma not_mem_bot (x : ℕ × ℕ) : x ∉ (⊥ : young_diagram) := finset.not_mem_empty x
+
+instance : inhabited young_diagram := ⟨⊥⟩
+
+instance : distrib_lattice young_diagram :=
+function.injective.distrib_lattice
+  young_diagram.cells
+  (λ μ ν h, by rwa young_diagram.ext_iff)
+  (λ _ _, rfl) (λ _ _, rfl)
+
+end distrib_lattice
+
+/-- Cardinality of a Young diagram -/
+@[reducible] protected def card (μ : young_diagram) : ℕ := μ.cells.card
+lemma card_def (μ : young_diagram) : μ.card = μ.cells.card := rfl
```

Comment:
Arguably you don't need this lemma either, it's generated automatically as an equation lemma, and `rw card_def` is therefore the same as `rw card`.